library/alloc/src/rc.rs

```diff
@@ -2512,14 +2512,23 @@ trait RcInnerPtr {
     fn inc_strong(&self) {
         let strong = self.strong();
 
+        // We insert an `assume` here to hint LLVM at an otherwise
+        // missed optimization.
+        // SAFETY: The reference count will never be zero when this is
+        // called.
+        unsafe {
+            core::intrinsics::assume(strong != 0);
+        }
+
+        let strong = strong.wrapping_add(1);
+        self.strong_ref().set(strong);
```

Comment:
Consider moving the write to after the abort, so there's never a window where the reference count is incorrectly zero. (The process could be interrupted during such a window, and invalidate the above `assume`.)

Reply:
Putting the abort after the write was intentionally chosen because it produces the best machine code: https://rust.godbolt.org/z/P6nTf5K7j

> (The process could be interrupted during such a window, and invalidate the above `assume`.)

How? I don't think that's possible.

Reply:
I think this is okay. As far as I can tell @mjbshaw is correct and there is no way to observe the inconsistency. Given that and the concrete benefits with respect to codegen I'm happy to temporarily break that invariant in the case of an upcoming abort.